pkg/webhook/admission/deployment/model_provisioning_injector.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2019 kubeflow.org.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+	http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployment
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/constants"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	ModelProvisioningContainerName    = "model-provisioner"
+	ModelProvisioningVolumeName       = "kfserving-provision-location"
+	ModelProvisioningContainerImage   = "kcorer/kfdownloader"
+	ModelProvisioningContainerVersion = "latest"
+	PvcURIPrefix                      = "pvc://"
+	PvcSourceMountName                = "kfserving-pvc-source"
+	PvcSourceMountPath                = "/mnt/pvc"
+)
+
+// InjectModelProvisioner injects an init container to provision model data
+// for the serving container in a unified way across storage tech by injecting
+// a provisioning INIT container. This is a work around because KNative does not
+// support INIT containers: https://github.com/knative/serving/issues/4307
+func InjectModelProvisioner(deployment *appsv1.Deployment) error {
+
+	var srcURI, mountPath string
+
+	annotations := deployment.Spec.Template.ObjectMeta.Annotations
+	podSpec := &deployment.Spec.Template.Spec
+
+	// Only inject if the required annotations are set
+	if _, ok := annotations[constants.KFServiceModelProvisioningSourceURIAnnotationKey]; ok {
+		srcURI = annotations[constants.KFServiceModelProvisioningSourceURIAnnotationKey]
+	} else {
+		return nil
+	}
+
+	if _, ok := annotations[constants.KFServiceModelProvisioningMountPathAnnotationKey]; ok {
+		mountPath = annotations[constants.KFServiceModelProvisioningMountPathAnnotationKey]
+	} else {
+		mountPath = constants.DefaultModelLocalMountPath
+	}
+
+	// Only inject provisioning for supported URIs
+	if !strings.HasPrefix(srcURI, "gs://") &&
+		!strings.HasPrefix(srcURI, "s3://") &&
+		!strings.HasPrefix(srcURI, "pvc://") {
+		// TODO: would be nice to log something here so that future generations know what happened?
+		return nil
+	}
+
+	// Find the knative user-container (this is the model inference server)
+	userContainerIndex := -1
+	for idx, container := range podSpec.Containers {
+		if strings.Compare(container.Name, "user-container") == 0 {
+			userContainerIndex = idx
+			break
+		}
+	}
+
+	if userContainerIndex < 0 {
+		return fmt.Errorf("aargh no conuser container: %s", srcURI)
```

Comment:
in this negative case, we can ignore it, taking it as the non-knative-generated deployment. Maybe a warning message is enough

Reply:
The annotation is private - therefore if it has reached this point it is an error